ci: validate the message of each new commit as part of the CI linting

IgorMinar opened the pull request.

vicb commented on tools/validate-commit-message/index.js:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  const scope = match[2];
+  if (scope && scope.length > 0 && config['scopes'].indexOf(scope) === -1) {
```
`scope.length > 0` useless

vicb replied:
indexOf === -1 -> .includes ?

vicb commented on tools/validate-commit-message/index.js:
```diff
@@ -0,0 +1,50 @@
+ *
+ * Installation:
+ * >> cd <repo>
+ * >> ln -s ../../lib/validate-commit-msg.js .git/hooks/commit-msg
```
If this is intended to be a hook, we should split the headline vs the body (we do not want to restrict the body length)

vicb commented on gulpfile.js:
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
+    }
+
+    let someCommitsInvalid = false;
+    let commitsByLine = stdout.trim().split(/\n/);
```
trim ?

vicb commented on gulpfile.js:
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
+  const validateCommitMessage = require('./tools/validate-commit-message');
+  const childProcess = require('child_process');
+
+  childProcess.exec('git log --reverse --oneline HEAD ^upstream/master', (error, stdout, stderr) => {
```
`git log --reverse --oneline HEAD ^FETCH_HEAD`
actually `git log --reverse --oneline HEAD ^master` is probably good ?

vicb commented on gulpfile.js:
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
+
+    if (commitsByLine.length == 0) {
+      console.log('There are zero new commits between this HEAD and master');
+      process.exit(1);
```
1 ?

vicb commented on gulpfile.js:
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
+
+    commitsByLine.forEach(line => {
+      // the oneline format per line is: "SHA MESSAGE"
+      const firstSpace = line.indexOf(' ');
```
do not request the SHA if we do not use it ? `git log --reverse HEAD ^master --format=%s`

vicb commented on tools/validate-commit-message/index.js:
```diff
@@ -0,0 +1,50 @@
+const config = JSON.parse(fs.readFileSync(configPath, "utf8"));
+const PATTERN = /^(\w+)(?:\(([\w\$\.\*/-]*)\))?\: (.+)$/;
+
+module.exports = function(commitMessage) {
```
commitMessage -> subject

vicb commented on tools/validate-commit-message/index.js:
```diff
@@ -0,0 +1,50 @@
+const path = require('path');
+const configPath = path.resolve(__dirname, './commit-message.json');
+const config = JSON.parse(fs.readFileSync(configPath, "utf8"));
+const PATTERN = /^(\w+)(?:\(([\w\$\.\*/-]*)\))?\: (.+)$/;
```
`[\w\$\.\*/-]*` -> `[\w-]+` ?

vicb replied:
IIUC currently `type(): ...` would be a false positive

IgorMinar replied:
fixed. good catch

vicb commented on tools/validate-commit-message/index.js:
```diff
@@ -0,0 +1,50 @@
+
+  const type = match[1];
+  if (config['types'].indexOf(type) === -1) {
+    error(`${type} is not an allowed type.\n => TYPES: ${config['types'].join('|')}`, commitMessage);
```
join(', ') for readability ?

IgorMinar replied:
done

IgorMinar commented:
overriding CLA because @matsko signed it and I'm just taking over his PR with his permission.

IgorMinar committed:
ci: validate the message of each new commit as part of the CI linting

This patch adds the gulp command of `validate-commit-messages`
which will validate the range of commits messages present in the
active branch.

This check now runs on CI as part of the linting checks.

Allowed commit message types and scopes are controlled via commit-message.json file
and documented at https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines

This solution is based on old Vojta's code that he wrote for angular/angular.js, that was later adjusted
by @matsko in #13815.

Ideally we should switch over to something like https://www.npmjs.com/package/commitplease
as suggested in #9953 but that package currently doesn't support strict scope checking,
which is one of the primarily goal of this PR.

Note that this PR removes support for "chore" which was previously overused
by everyone on the team.

Closes #13815
Fixes #3337

IgorMinar commented:
@vicb I've cleaned everything up, added docs and tests. it should be good to go.

in the end I didn't use `git log --reverse --oneline HEAD ^FETCH_HEAD` because I think that's a bit unpredictable on local machine where what you last fetched might not be the last thing that's in upstream or origin. diffing against origin seems to work just fine.

IgorMinar committed:
ci: bump node and npm versions in circle.yaml to match travis

gkalpak commented on CONTRIBUTING.md:
```diff
@@ -191,21 +191,46 @@ If the commit reverts a previous commit, it should begin with `revert: `, follow
+* none/empty string: useful for `style`, `test` and `refactor` changes that are done across all packages (e.g. `style: add missing semicolons`)
+
+
+Packagigng
```
Typo?

IgorMinar replied:
yes. fixed

gkalpak replied:
I should have been clearer: I meant that the whole word should probably be removed.

gkalpak commented on gulpfile.js:
```diff
@@ -155,6 +155,42 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
+          if (!validateCommitMessage(commitMessage)) {
+            someCommitsInvalid = true;
+          }
+        });
```
`someCommitsInvalid = !commitsByLine.every(validateCommitMessage)` is shorter, more readable and efficient 😛

IgorMinar replied:
done. thanks

vicb commented on CONTRIBUTING.md:
```diff
@@ -191,21 +191,46 @@ If the commit reverts a previous commit, it should begin with `revert: `, follow
+* **common**
+* **compiler**
+* **compiler-cli**
+* **core**
```
What about `i18n` ? Should we allow `compiler/i18n` ? 
Not blocking the PR but let's discuss this

IgorMinar replied:
I'm implementing the current conventions that match the current changelog output. we can discuss can discuss bigger changes to this conventions in a separate commit.

vicb commented on gulpfile.js:
```diff
@@ -155,6 +155,40 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
+  // We need to fetch origin explicitly because it might be stale.
+  // I couldn't find a reliable way to do this without fetch.
+  childProcess.exec(
+      'git fetch origin master && git log --reverse --format=%s HEAD ^origin/master',
```
couldn't we use the local `master`:
- should be the same as `origin/master` on the CI,
-should be at least the fork point on your local dev

IgorMinar replied:
I tried but on Circle CI the master is almost always stale for some reason. this was the only way I could get reliably results.

vicb commented on gulpfile.js:
```diff
@@ -155,6 +155,40 @@ gulp.task('lint', ['format:enforce', 'tools:build'], () => {
+
+        console.log(`Examining ${commitsByLine.length} commits between HEAD and master`);
+
+        if (commitsByLine.length == 0) {
```
```
if (0) exit(0)
else
   ...
```

vicb commented on tools/validate-commit-message/validate-commit-message.js:
```diff
@@ -0,0 +1,58 @@
+ *
+ * Installation:
+ * >> cd <repo>
+ * >> ln -s ../../tools/validate-commit-message/validate-commit-message.js .git/hooks/commit-msg
```
not addressed

IgorMinar replied:
hmm.. what was the issue with this? I'll look up your old comment.

IgorMinar replied:
> If this is intended to be a hook, we should split the headline vs the body (we do not want to restrict the body length)

hmm. you are right. originally when Vojta wrote it was meant to be a commit hook, but with all the changes it turned into something else.

I'll change the code to check just the first line.

IgorMinar replied:
actually that would be a more complicated change. I'll just drop the comment.

vicb commented on tools/validate-commit-message/validate-commit-message.js:
```diff
@@ -0,0 +1,58 @@
+const path = require('path');
+const configPath = path.resolve(__dirname, './commit-message.json');
+const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
+const PATTERN = /^(revert\: )?(\w+)(?:\(([\w\$\.\*/-]*)\))?\: (.+)$/;
```
previous comment not addressed

vicb commented on tools/validate-commit-message/validate-commit-message.js:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  const match = PATTERN.exec(commitMessage);
+  if (!match || match[2] === 'revert') {
```
why is revert special here ?

IgorMinar replied:
it's always been this way. we just never enforced it. the revert should contain the whole original message so it's not a special type. it's more like a prefix for the usual message format.

vicb commented on tools/validate-commit-message/validate-commit-message.spec.js:
```diff
@@ -0,0 +1,117 @@
+
+
+    it('should fail when scope is invalid', function() {
+      expect(validateMessage('fix(Compiler): something')).toBe(INVALID);
```
to we want to be case insensitive ?

IgorMinar replied:
the packages are all lowercase. enforcing the case is a good thing for consistency.

vicb commented on tools/validate-commit-message/validate-commit-message.js:
```diff
@@ -0,0 +1,58 @@
+
+  if (scope && !config['scopes'].includes(scope)) {
+    error(
+        `"${scope}" inside of "${type}(...)" is not an allowed scope.\n => SCOPES: ${config['scopes'].join('|')}`,
```
remove or rephrase `inside of "${type}(...)"` when looking at the error msgs in the tests, what I read is that I can use whatever scope but not inside the current type vs the scope is invalid.

vicb commented:
LGTM, needs a green CI (cleanup)
